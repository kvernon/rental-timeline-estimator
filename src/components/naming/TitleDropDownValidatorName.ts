import { formatName } from './FormatName';

export const TitleDropDownValidatorId = 'titleDropDown';
export const TitleDropDownValidatorName = (id: string) => formatName(id, TitleDropDownValidatorId);
