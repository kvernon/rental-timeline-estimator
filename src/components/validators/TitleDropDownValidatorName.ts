export const TitleDropDownValidatorName = (id: string) => `titleDropDown${id}`;
