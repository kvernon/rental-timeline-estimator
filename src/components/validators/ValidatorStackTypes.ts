export enum ValidatorStackTypes {
  Optional,
  Required,
}
