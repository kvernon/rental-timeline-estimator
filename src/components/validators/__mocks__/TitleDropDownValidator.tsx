import type { ITitleDropDownParams } from '../TitleDropDownValidator';
import React from 'react';
import { ValidatorTypes } from '../ValidatorTypes';

export const TitleDropDownValidator = jest.fn((p: ITitleDropDownParams) => {
  function handleChange(event: { currentTarget: HTMLSelectElement }): void {
    const value = parseInt(event.currentTarget.value);

    if (p.onChange) {
      p.onChange({ value: { value, label: '' }, validationResult: ValidatorTypes.Valid });
    }
  }

  return (
    <select value={p.value?.value?.value} onChange={handleChange} aria-label={p.title}>
      {p.optionTitles.map((title, index) => (
        <option key={title} value={index}>
          {title}
        </option>
      ))}
    </select>
  );
});
