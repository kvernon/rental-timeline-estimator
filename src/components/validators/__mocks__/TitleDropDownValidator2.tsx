import type { ITitleDropDownParams } from '../TitleDropDownValidator2';
import React from 'react';

export const TitleDropDownValidator2 = jest.fn((p: ITitleDropDownParams) => {
  function handleChange(event: { currentTarget: HTMLSelectElement }): void {
    const value = parseInt(event.currentTarget.value);
    const option = { label: p.titles[value], value };

    if (p.onChange) {
      p.onChange(option);
    }
  }

  return (
    <select id={`titleDropDownValidator${p.id}`} value={p.defaultIndex} onChange={handleChange}>
      {p.titles.map((title, index) => (
        <option key={title} value={index}>
          {title}
        </option>
      ))}
    </select>
  );
});
