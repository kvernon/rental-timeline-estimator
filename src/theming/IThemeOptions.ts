import { IPalette } from './IPalette';

import { ITypography } from './ITypography';

export interface IThemeOptions {
  palette: IPalette;
  typography: ITypography;
}
