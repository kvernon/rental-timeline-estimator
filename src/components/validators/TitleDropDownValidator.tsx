import React, { useEffect, useState } from 'react';
import ReactSelect, { SingleValue } from 'react-select';
import { useTheme } from '@emotion/react';
import { IThemeOptions } from '../../theming/IThemeOptions';
import { evaluateValidation, RuleEval } from './evaluatateValidation';
import { ValidatorTypes } from './ValidatorTypes';
import { FontGroups } from '../../theming/fontGroups';
import { ValidatorStackTypes } from './ValidatorStackTypes';
import styled from '@emotion/styled';

export interface ITitleDropDownParams {
  titles: string[];
  defaultIndex?: number;
  id?: string;
  validationType: ValidatorStackTypes;
  onChange?: (value: ITitleDropDownOption) => void;
}

export interface ITitleDropDownOption {
  label: string;
  value: number;
}

const rule: RuleEval = (v: number, options: { min?: number; max?: number }) =>
  v > (options?.min || 0) ? ValidatorTypes.Valid : ValidatorTypes.Invalid;

export const TitleDropDownValidator = function (props: ITitleDropDownParams) {
  const coreTheme = useTheme() as IThemeOptions;

  const selectUuid = props.id || window.crypto.randomUUID();

  const [optionsMap, setOptionsMap] = useState<ITitleDropDownOption[]>([]);

  const [selectedIndex, setSelectedIndex] = useState<number>(props.defaultIndex || 0);
  const [evaluated, setEvaluated] = useState(evaluateValidation(props.validationType, rule, selectedIndex));

  useEffect(() => {
    const map = props.titles.map((title: string, idx: number): ITitleDropDownOption => {
      return { value: idx, label: title };
    });

    if (JSON.stringify(optionsMap) !== JSON.stringify(map)) {
      setOptionsMap(map);
      //console.log('here', map);
    }
  }, [props.titles, optionsMap]);

  useEffect(() => {
    const eventResult = evaluateValidation(props.validationType, rule, selectedIndex);

    if (evaluated.validationResult !== eventResult.validationResult) {
      setEvaluated(eventResult);
    }
  }, [evaluated, props.validationType, selectedIndex]);

  const handleChange = (option: SingleValue<ITitleDropDownOption | unknown>): void => {
    const value = (option as ITitleDropDownOption).value;
    if (value !== selectedIndex) {
      setSelectedIndex(value);
      if (props.onChange) {
        props.onChange({ value, label: optionsMap[value].label });
      }
    }
  };

  const Select = styled(ReactSelect)<{
    themeOptions: IThemeOptions;
  }>`
    appearance: none;
    white-space: pre-wrap;
    font-size: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.size};
    font-family: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.font};
    font-weight: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.weight};
    width: 100%;
    padding-left: 10px;
    color: ${(props) => props.themeOptions.typography.get(FontGroups.input)?.color};
    overflow: visible;
  `;

  return (
    <Select
      {...props}
      name={selectUuid}
      isMulti={false}
      value={optionsMap[selectedIndex]}
      themeOptions={coreTheme}
      options={optionsMap}
      onChange={handleChange}
      styles={{
        singleValue: (base) => ({
          ...base,
          color: `${coreTheme.typography.get(FontGroups.input)?.color}`,
        }),
        menu: (baseStyles) => ({
          ...baseStyles,
          zIndex: baseStyles.isSelected ? 9999 : baseStyles.zIndex,
        }),
        control: (baseStyles) => {
          return {
            ...baseStyles,
            overflow: 'visible',
            transition: 'background-color 0.4s ease-out',
            backgroundColor: `${coreTheme.palette.validation[evaluated.validationResultName].background}41`,
            height: '59px',
            borderColor: `${coreTheme.palette.inputBackground}`,
            border: `1px solid ${coreTheme.palette.panelBackground}`,
            borderRadius: '0.3rem',
            color: `${coreTheme.typography.get(FontGroups.input)?.color}`,
            ':hover': {
              backgroundColor: `${coreTheme.palette.validation[evaluated.validationResultName].background}81`,
              borderColor: `${coreTheme.palette.inputBackgroundFocus}`,
              color: `${coreTheme.palette.inputBackgroundFocus}`,
            },
          };
        },
      }}
    />
  );
};
