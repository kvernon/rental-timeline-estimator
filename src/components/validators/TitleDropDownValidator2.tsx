import { IOption, ITitleDropDownParams } from './TitleDropDownValidator';
import React, { useCallback, useEffect, useState } from 'react';
import { TitleDropDownValidatorName } from '../naming/TitleDropDownValidatorName';
import { Controller, useFormContext } from 'react-hook-form';
import ReactSelect, { SingleValue } from 'react-select';
import { useTheme } from '@emotion/react';
import { IThemeOptions } from '../../theming/IThemeOptions';
import { evaluateValidation, RuleEval } from './evaluatateValidation';
import { ValidatorTypes } from './ValidatorTypes';
import styled from '@emotion/styled';
import { useWatcher } from '../hooks/useWatcher';
import { FontGroups } from '../../theming/fontGroups';

const rule: RuleEval = (v: number, options: { min?: number; max?: number }) =>
  v > (options?.min || 0) ? ValidatorTypes.Valid : ValidatorTypes.Invalid;

export const TitleDropDownValidator2 = function (props: ITitleDropDownParams) {
  const { control, setValue } = useFormContext();
  const coreTheme = useTheme() as IThemeOptions;

  const selectUuid = props.id || window.crypto.randomUUID();
  const id = `${TitleDropDownValidatorName(selectUuid)}.value`;
  const inputValidationResult = `${TitleDropDownValidatorName(selectUuid)}.validationResult`;

  const [optionsMap, setOptionsMap] = useState<IOption[]>([]);

  const [selectedIndex, setSelectedIndex] = useState<number>(props.defaultIndex || 0);
  const [evaluated, setEvaluated] = useState(evaluateValidation(props.validationType, rule, selectedIndex));

  const [watcherResult] = useWatcher<IOption>([id]);

  useEffect(() => {
    const map = props.titles.map((title: string, idx: number): IOption => {
      return { value: idx, label: title };
    });

    if (JSON.stringify(optionsMap) !== JSON.stringify(map)) {
      setOptionsMap(map);
    }
  }, [props.titles, optionsMap]);

  useEffect(() => {
    const eventResult = evaluateValidation(props.validationType, rule, selectedIndex);

    if (evaluated.validationResult !== eventResult.validationResult) {
      setEvaluated(eventResult);
      setValue(inputValidationResult, evaluated.validationResult);
    }
  }, [watcherResult, evaluated, inputValidationResult, props.validationType, setValue, selectedIndex]);

  const Select = styled(ReactSelect)<{ themeOptions: IThemeOptions }>`
    appearance: none;
    white-space: pre-wrap;
    font-size: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.size};
    font-family: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.font};
    font-weight: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.weight};
    width: 100%;
    padding-left: 10px;
    color: ${(props) => props.themeOptions.typography.get(FontGroups.input)?.color};
    overflow: visible;
  `;

  const handleChange = useCallback(
    (option: SingleValue<IOption | unknown>): void => {
      const value = (option as IOption).value;
      if (value !== selectedIndex) {
        setSelectedIndex(value);
      }
    },
    [selectedIndex, id, setSelectedIndex],
  );

  return (
    <Controller
      name={id}
      control={control}
      render={({ field }) => {
        return <Select {...field} isMulti={false} themeOptions={coreTheme} options={optionsMap} onChange={handleChange} />;
      }}
    />
  );
};
