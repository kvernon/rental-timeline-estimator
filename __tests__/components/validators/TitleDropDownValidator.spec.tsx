import { Theme, useTheme } from '@emotion/react';

import { configure, render, screen } from '@testing-library/react';
import React from 'react';
import { ITitleDropDownParams } from '../../../src/components/validators/TitleDropDownValidator';
import { ValidatorStackTypes } from '../../../src/components/validators/ValidatorStackTypes';
import { TitleDropDownValidatorName } from '../../../src/components/naming/TitleDropDownValidatorName';
import { TitleDropDownValidator2 } from '../../../src/components/validators/TitleDropDownValidator2';
import { FormProvider, useForm } from 'react-hook-form';
import { IThemeOptions } from '../../../src/theming/IThemeOptions';
import { ITypography } from '../../../src/theming/ITypography';

const Setup = (props: ITitleDropDownParams) => {
  const methods = useForm({ mode: 'onBlur' });
  return (
    <FormProvider {...methods}>
      <TitleDropDownValidator2 {...props} />
    </FormProvider>
  );
};

jest.mock('@emotion/react', () => {
  const requireActual = jest.requireActual('@emotion/react');
  const useTheme: jest.MockedFn<() => Theme> = jest.fn();
  return {
    ...requireActual,
    useTheme,
  };
});
describe.only('TitleDropDownValidator unit tests', () => {
  let typographyMock: jest.Mocked<ITypography>;

  const validationColorOptionalRight = '#0000FF';
  const validationColorValidMiddle = '#00FF00';
  const validationColorInvalidLeft = '#FF0000';

  afterEach(() => {
    jest.clearAllMocks();
  });

  const p: ITitleDropDownParams = {
    id: 'Tested',
    titles: [],
    validationType: ValidatorStackTypes.Required,
  };

  beforeEach(() => {
    configure({ testIdAttribute: 'name' });
    const useThemeMock = jest.mocked(useTheme);
    typographyMock = {
      parent: {
        allPopulated: jest.fn(),
        font: 'p',
        color: 'p',
        size: 'p',
      },
      get: jest.fn().mockReturnValue({
        font: 'child',
        color: 'child',
        size: 'child',
      }),
    };

    useThemeMock.mockReturnValue({
      palette: {
        pageBackground: 'pageBackground',
        panelBackground: 'panelBackground',
        panelShadow: 'panelShadow',

        inputBackgroundBadFocus: 'inputBackgroundBadFocus',
        inputBackgroundBad: 'inputBackgroundBad',
        inputBackground: 'inputBackground',
        inputBackgroundFocus: 'inputBackgroundFocus',

        validation: {
          Invalid: {
            validationColor: validationColorInvalidLeft,
            background: validationColorInvalidLeft,
            backgroundFocus: validationColorInvalidLeft,
          },
          Valid: {
            validationColor: validationColorValidMiddle,
            background: validationColorValidMiddle,
            backgroundFocus: validationColorValidMiddle,
          },
          Optional: {
            validationColor: validationColorOptionalRight,
            background: validationColorOptionalRight,
            backgroundFocus: validationColorOptionalRight,
          },
        },
      },
      typography: typographyMock,
    } as jest.Mocked<IThemeOptions>);

    render(<Setup {...p} />);
  });

  test('should contain styles', () => {
    const entity = screen.getByTestId<HTMLInputElement>(`${TitleDropDownValidatorName(p.id as string)}.value`);

    expect(entity).toMatchSnapshot();
  });
});
