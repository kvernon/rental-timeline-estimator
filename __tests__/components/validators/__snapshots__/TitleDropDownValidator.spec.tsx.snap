// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleDropDownValidator unit tests should contain styles 1`] = `
<input
  name="titleDropDownTested.value"
  type="hidden"
  value=""
/>
`;
